cpp/src/decisiontree/decisiontree.cu

```diff
@@ -105,246 +209,360 @@ void DecisionTreeParams::print() const {
 	}
-	
-	std::random_shuffle(feature_selector.begin(),feature_selector.end());
-	feature_selector.resize((int) (colper * dinfo.Ncols));
-	
+
+	std::random_shuffle(this->feature_selector.begin(), this->feature_selector.end());
+	this->feature_selector.resize((int) (colper * this->dinfo.Ncols));
+
 	cudaDeviceProp prop;
 	CUDA_CHECK(cudaGetDeviceProperties(&prop, 0));
-	max_shared_mem = prop.sharedMemPerBlock;
+	this->max_shared_mem = prop.sharedMemPerBlock;
 
-	if (split_algo == SPLIT_ALGO::HIST) {
-		shmem_used += 2 * sizeof(T) * ncols;
-		shmem_used += nbins * n_unique_labels * sizeof(int) * ncols;
+	if (this->split_algo == SPLIT_ALGO::HIST) {
+		this->shmem_used += 2 * sizeof(T) * ncols;
+		this->shmem_used += this->nbins * this->n_unique_labels * sizeof(int) * ncols;
 	} else {
-		shmem_used += nbins * n_unique_labels * sizeof(int) * ncols;
+		this->shmem_used += this->nbins * this->n_unique_labels * sizeof(int) * ncols;
 	}
-	ASSERT(shmem_used <= max_shared_mem, "Shared memory per block limit %zd , requested %zd \n", max_shared_mem, shmem_used);
+	ASSERT(this->shmem_used <= this->max_shared_mem, "Shared memory per block limit %zd , requested %zd \n", this->max_shared_mem, this->shmem_used);
 
-	for (int i = 0; i < MAXSTREAMS; i++) {
-		tempmem[i] = std::make_shared<TemporaryMemory<T>>(handle, n_sampled_rows, ncols, MAXSTREAMS, unique_labels, n_bins, split_algo);
-		if (split_algo == SPLIT_ALGO::GLOBAL_QUANTILE) {
-			preprocess_quantile(data, rowids, n_sampled_rows, ncols, dinfo.NLocalrows, n_bins, tempmem[i]);
+	for (int i = 0; i < this->MAXSTREAMS; i++) {
+		this->tempmem[i] = std::make_shared<TemporaryMemory<T, int>>(handle, n_sampled_rows, ncols, this->MAXSTREAMS, unique_labels, n_bins, this->split_algo);
+		if (this->split_algo == SPLIT_ALGO::GLOBAL_QUANTILE) {
+			preprocess_quantile(data, rowids, n_sampled_rows, ncols, this->dinfo.NLocalrows, n_bins, this->tempmem[i]);
 		}
 	}
-	total_temp_mem = tempmem[0]->totalmem;
-	total_temp_mem *= MAXSTREAMS;
-	GiniInfo split_info;
+	this->total_temp_mem = this->tempmem[0]->totalmem;
+	this->total_temp_mem *= this->MAXSTREAMS;
+	MetricInfo<T> split_info;
 	MLCommon::TimerCPU timer;
-	root = grow_tree(data, colper, labels, 0, rowids, n_sampled_rows, split_info);
-	construct_time = timer.getElapsedSeconds();
+	this->root = grow_tree(data, colper, labels, 0, rowids, n_sampled_rows, split_info);
+	this->construct_time = timer.getElapsedSeconds();
 
-	for (int i = 0; i < MAXSTREAMS; i++) {
-		tempmem[i].reset();
+	for (int i = 0; i < this->MAXSTREAMS; i++) {
+		this->tempmem[i].reset();
 	}
 
 	return;
 }
 
 template<typename T>
-TreeNode<T>* DecisionTreeClassifier<T>::grow_tree(T *data, const float colper, int *labels, int depth, unsigned int* rowids,
-												const int n_sampled_rows, GiniInfo prev_split_info) {
+TreeNode<T, int>* DecisionTreeClassifier<T>::grow_tree(T *data, const float colper, int *labels, int depth, unsigned int* rowids,
```

Comment:
I liked the base-class design to reduce code duplication between classifier and regressor children! However, there still seems to be quite a bit of duplication between the child classes, eg: bootstrapping, and so on. Maybe all these commonalities could further to put inside the base class?

Reply:
Yes! Just pulled the plant and grow_tree methods to the base class (DecisionTreeBase).

Reply:
Thanks @myrtw for the changes